src/tui/mod.rs

```
@@ -0,0 +1,144 @@
+        .into_diagnostic()
+        .wrap_err("Failed to capture stderr")?;
+
+    let mut ghci_buffer = [0; 1024];
+    let mut tracing_buffer = [0; 1024];
+
+    let mut terminal = terminal::enter()?;
+
+    let mut tui = Tui::default();
+
+    let mut event_stream = EventStream::new();
+
+    while !tui.quit {
+        let mut render_result = Ok(());
+        terminal
+            .draw(|frame| {
+                let area = frame.size();
+                let buffer = frame.buffer_mut();
+                render_result = render(&tui, area, buffer);
+            })
+            .into_diagnostic()
+            .wrap_err("Failed to draw to terminal")?;
+        render_result?;
+
+        tokio::select! {
+            _ = shutdown.on_shutdown_requested() => {
+                tui.quit = true;
+            }
+
+            output = ghci_reader.read(&mut ghci_buffer) => {
+                let n = output
+                    .into_diagnostic()
+                    .wrap_err("Failed to read bytes from GHCi into TUI buffer")?;
+                if n == 0 {
+                    tui.quit = true;
+                } else {
+                    tui.scrollback.extend(ghci_buffer);
+                    ghci_buffer = [0; 1024];
+                }
+            }
+
+            output = tracing_reader.read(&mut tracing_buffer) => {
+                output
+                    .into_diagnostic()
+                    .wrap_err("Failed to read bytes from tracing into TUI buffer")?;
+                tui.scrollback.extend(tracing_buffer);
+                tracing_buffer = [0; 1024];
+            }
+
+            output = event_stream.next() => {
+                let event = output
+                    .ok_or_else(|| miette!("No more crossterm events"))?
+                    .into_diagnostic()
+                    .wrap_err("Failed to get next crossterm event")?;
+                handle_event(&mut tui, event)?;
+            }
+        }
+    }
+
+    let _ = shutdown.request_shutdown();
```

Ignoring any errors because that's what's done everywhere else with `request_shutdown`, and I think a possible error is "somebody else already requested shutdown"?